ComPact/Builders/V3/MessageBuilder.cs

```diff
@@ -106,11 +146,38 @@ private void CheckMessageHandler<T>(Action<T> messageHandler)
             {
                 throw new ArgumentNullException(nameof(messageHandler));
             }
+            CheckMessageContents();
+            _isVerified = false;
+        }
+
+        private void CheckMessageHandler<T>(Func<T,Task> messageHandler)
+        {
+            if (messageHandler is null)
+            {
+                throw new ArgumentNullException(nameof(messageHandler));
+            }
+            CheckMessageContents();
+            _isVerified = false;
+        }
+
+        private void CheckMessageContents()
+        {
             if (_message.Contents == null)
             {
                 throw new PactException("Message content has not been set. Please provide using the With method.");
+            }
+        }
+        private async Task InvokeMessageHandlerAsync<T>(Func<T,Task> messageHandler, T content)
+        {
+            try
+            {
+                await messageHandler.Invoke(content);
             }
-            _isVerified = false;
+            catch(Exception exception)
+            {
+                throw new PactException("Message handler threw an exception",exception);
```

Comment:
Feel free to add this to the other InvokeMessageHandler method as well.

Reply:
updated all exception at file